src/gsi/cplr_wrwrfmassa.f90

```diff
@@ -2200,18 +2261,158 @@ subroutine wrwrfmassa_netcdf_wrf(this,mype)
+                            qg_tmp=zero
+                         end if
+
+                      else if ( l_use_cvpqx_pval .gt. 0. ) then   ! CVpq
+!                        convert log(qr/qs/qg) back to qr/qs/qg
+                         if (mype==0 .AND. i==3 .AND. j==3 .AND. k==3) then
+                            write(6,*)'wrwrfmass_netcdf: convert power transformed(qr/qs/qg) back to qr/qs/qg.'
+                         end if
+
+                         qr_tmp=max((l_use_cvpqx_pval*ges_qr(j,i,k)+1)**(1/l_use_cvpqx_pval)-1.0E-6_r_kind,0.0)
+                         qs_tmp=max((l_use_cvpqx_pval*ges_qs(j,i,k)+1)**(1/l_use_cvpqx_pval)-1.0E-6_r_kind,0.0)
+                         qg_tmp=max((l_use_cvpqx_pval*ges_qg(j,i,k)+1)**(1/l_use_cvpqx_pval)-1.0E-6_r_kind,0.0)
+
+                         !Rong Kong add a upper limit to hydrometeors to disable overshooting
+                         qr_tmp=min(qr_tmp,1E-2_r_kind)
+                         qs_tmp=min(qs_tmp,1.5E-2_r_kind)
+                         qg_tmp=min(qg_tmp,2E-2_r_kind)
+                         !Rong Kong added above
+                      end if
+                   else   ! For CVq
+
+                      qr_min=zero
+                      qs_min=zero
+                      qg_min=zero
+
+                      if (mype==0 .AND. i==3 .AND. j==3 .AND. k==3) then
+                         write(6,*)'wrwrfmass_netcdf: only reset (qr/qs/qg) to 0.0 for negative analysis value. (regular qx)'
+                      end if
+
+                      qr_tmp=ges_qr(j,i,k)-1.0E-8_r_kind
+                      qs_tmp=ges_qs(j,i,k)-1.0E-8_r_kind
+                      qg_tmp=ges_qg(j,i,k)-1.0E-8_r_kind
+
+
+                      qr_tmp=max(qr_tmp,qr_min)
+                      qs_tmp=max(qs_tmp,qs_min)
+                      qg_tmp=max(qg_tmp,qg_min)
+
+                   end if         ! cvpqx
+
+                   all_loc(j,i,kqr)=qr_tmp
+                   all_loc(j,i,kqs)=qs_tmp
+                   all_loc(j,i,kqg)=qg_tmp
+
+!             =========================================================================
+!             !
+!             re-initialisation of QNRAIN with analyzed qr and N0r(which is  single-moment parameter) 
+!             equation is used in subroutine init_MM of initlib3d.f90 in arps pacakge
+                   qnr_tmp = zero
+                   if ( cld_nt_updt .eq. 2 ) then
+                      T1D=ges_tsen(j,i,k,it) ! sensible temperature (K)
+                      P1D=r100*(aeta1_ll(k)*(r10*ges_ps(j,i)-pt_ll)+pt_ll) ! pressure hPa --> Pa
+                      Q1D=ges_q(j,i,k)/(one-ges_q(j,i,k)) ! mixing ratio
+                      RHO=P1D/(rd*T1D*(one+D608*Q1D)) ! air density in kg m^-3
+                      QR1D=qr_tmp
+                      CALL init_mm_qnr(RHO,QR1D,qnr_tmp)
+                      qnr_tmp = max(qnr_tmp, zero)
+
+                      if (mype==0 .AND. i==25 .AND. j==25 .AND. k==15) then
```

Comment:
convert this kind of write to comments or move them out of the loop.

Reply:
I modified the code according to your comments. I took out these codes from loops and make them print one time at the start of the routine using if(mype==0).

https://github.com/NOAA-EMC/GSI/pull/142/files#diff-83013677e0fb286b5f115ca5766f4b67266ed22b3d70b8867ed865db67ab8f50R3398-R3400